frontend/src/views/instructor/ddahs/import.tsx

```diff
@@ -0,0 +1,319 @@
+import React from "react";
+import { Alert } from "react-bootstrap";
+import { FaArrowRight } from "react-icons/fa";
+import { useSelector } from "react-redux";
+import { assignmentsSelector, applicantsSelector } from "../../../api/actions";
+import { ddahsSelector, upsertDdahs } from "../../../api/actions/ddahs";
+import {
+    Ddah,
+    Assignment,
+    Applicant,
+    MinimalDdah,
+    RawDuty,
+} from "../../../api/defs/types";
+import { ImportActionButton } from "../../../components/import-button";
+import { DiffSpec, diffImport, getChanged } from "../../../libs/diffs";
+import { normalizeDdahImports } from "../../../libs/import-export";
+import { useThunkDispatch } from "../../../libs/thunk-dispatch";
+
+enum DiffType {
+    Unchanged = "UNCHANGED",
+    Created = "CREATED",
+    Updated = "UPDATED",
+    Deleted = "DELETED",
+}
+
+type MinimalDuty = {
+    description: string;
+    hours: number;
+};
+
+interface DdahUpdate {
+    updatedDdah: Ddah;
+    oldDdah: Ddah;
+}
+
+interface DutyDiff {
+    oldDuty?: MinimalDuty;
+    newDuty?: MinimalDuty;
+    status: DiffType;
+}
+
+function getDutiesDiff(oldDuties: RawDuty[], newDuties: RawDuty[]) {
+    let changes: DutyDiff[] = [];
+    let dutiesChanged = false;
+
+    oldDuties.forEach((oldDuty) => {
+        let oldDutyDeleted = true;
+
+        newDuties.forEach((newDuty, index) => {
+            if (
+                oldDuty.description === newDuty.description &&
+                oldDuty.hours === newDuty.hours
+            ) {
+                changes.push({ oldDuty, newDuty, status: DiffType.Unchanged });
+                newDuties.splice(index, 1);
+                oldDutyDeleted = false;
+            } else if (
+                oldDuty.description === newDuty.description &&
+                oldDuty.hours !== newDuty.hours
+            ) {
+                changes.push({ oldDuty, newDuty, status: DiffType.Updated });
+                newDuties.splice(index, 1);
+                oldDutyDeleted = false;
+                dutiesChanged = true;
+            }
```

Comment:
I'm confused about what exactly is going on...it looks like this is the same as a map with two filters?

Reply:
I am not too sure I follow, which map and filters are you thinking of? :slightly_smiling_face:

Reply:
You're pushing and splicing...try to rewrite this as a map and see how it looks.

Reply:
I am not sure how to do that unfortunately, the `map` needs every old duty to be mapped to something that is not `undefined` (or else the types break) and for old duties that are deleted - they are "mapped" to `undefined`. That is why it's a loop, and not a map :)